skimage/transform/registration.py

```diff
@@ -0,0 +1,143 @@
+import numpy as np
+from scipy.optimize import basinhopping, minimize
+
+from .pyramids import pyramid_reduce
+from ._warps import warp, SimilarityTransform
+
+__all__ = ['register', 'p_to_matrix', 'matrix_to_p']
+
+
+def _gaussian_pyramid(image, levels=6):
+    """
+    Creates a gaussian pyramid for an image
+    
+    Parameters
+    ----------
+    image : (M, N) ndarray
+        Input image with is used as a base for the pyramid.
+    levels : int, optional (default 6)
+        Int equals to the amount of times the image is blurred and reduced
+    
+    Returns
+    -------
+    pyramid: array
+        Ordered array of images of size level starting with the most blurred
+        
+    """
+    pyramid = levels*[None]
+    pyramid[-1] = image
+
+    for level in range(levels-2, -1, -1):
+        image = pyramid_reduce(image, sigma=2/3)
+        pyramid[level] = (image)
+
+    return pyramid
+
+
+def _mse(img1, img2):
+    """
+    Finds the error between the overlapping area of two images
+    
+    Parameters
+    ----------
+    img1 : (M, N) ndarray
+        Input image used for reference
+    img2 : (M, N) ndarray
+        Input image which is compared to img1
+    
+    Returns
+    -------
+    err: int
+        Error in the form of the squared difference between pixels
+    """
+    dim0 = min(img1.shape[0],img2.shape[0])
+    dim1 = min(img1.shape[1],img2.shape[1])
+    return ((img1[:dim0,:dim1]-img2[:dim0,:dim1])**2).sum()
+
+
+def _cost_mse(param, reference_image, target_image):
+    """
+    Finds the error between the overlapping area of two images after transformations are done to one of them
+    
+    Parameters
+    ----------
+    param : array
+        Input array giving the translation and rotation the target image will undergo
+    reference_image : (M, N) ndarray
+        Input image used for reference
+    target_image : (M, N) ndarray
+        Input image which is modified and then compared to reference_image
+    
+    Returns
+    -------
+    err: int
+        Error in the form of the squared difference between pixels
+    """
+    transformation = p_to_matrix(param)
+    transformed = warp(target_image, transformation, order=3)
```

Comment:
We don't want to use `warp` here because it uses "backwards" coordinates. Use `ndi.affine_transform` instead